app/components/infiniteScroll.tsx

```diff
@@ -1,58 +1,43 @@
 import {ReactNode, useState, useRef, useCallback} from 'react'
-import {Question} from '~/server-utils/stampy'
 
 type Props = {
-  children: ReactNode[]
-  fetchMore: () => Promise<Question[]>
-  loadingRenderer?: (loading: boolean) => ReactNode
+  children: ReactNode
+  fetchMore: () => Promise<string | null>
 } & JSX.IntrinsicElements['div']
 
-export default function InfiniteScroll({
-  children,
-  fetchMore,
-  loadingRenderer = (loading: boolean) => loading && <div className="loader"></div>,
-  ...props
-}: Props) {
-  const [pageNum, setPageNum] = useState(0)
-  const [hasMore, setHasMore] = useState(true)
+export default function InfiniteScroll({children, fetchMore}: Props) {
   const [loading, setLoading] = useState(false)
 
-  const getMore = useCallback(async () => {
-    if (loading || !hasMore) return
-
-    setLoading(true)
-
-    const nextPage = await fetchMore()
-    setPageNum(pageNum + 1)
-
-    if (nextPage?.length === 0) {
-      setHasMore(false)
-    }
-
-    setLoading(false)
-  }, [loading, hasMore, pageNum, setLoading, setPageNum, setHasMore, fetchMore])
-
-  const observer = useRef<IntersectionObserver | null>(null)
-  const lastElementRef = useCallback(
-    (node) => {
-      if (observer.current) observer.current.disconnect()
-
-      if (node) {
-        observer.current = new IntersectionObserver((entries) => {
-          if (entries[0].isIntersecting) {
-            getMore()
-          }
-        })
-        observer.current.observe(node)
+  // must use stable callback without depencencies, otherwise it will create many instances of IntersectionObserver
+  const fetchMoreRef = useRef(fetchMore)
+  fetchMoreRef.current = fetchMore
+  const lastElementRef = useCallback((node: HTMLDivElement) => {
+    if (!node) return
+
+    let loadingInternal = false
+    const observer = new IntersectionObserver(async (entries) => {
+      if (entries[0].isIntersecting) {
+        // state is needed for re-render, internal variable to avoid dependencies
+        if (loadingInternal) return
+        setLoading(true)
+        loadingInternal = true
+
+        const nextPageLink = await fetchMoreRef.current()
+        if (!nextPageLink) {
+          observer.disconnect()
+        }
+        setLoading(false)
+        loadingInternal = false
       }
-    },
-    [getMore]
```

Comment:
@mruwnik since the dependencies changed almost in every rerender, it didn't make much sense to wrap it in `useCallback`, it was called from the callback ref quite a lot, disconnecting and reconnecting new observers .. not a big deal, but logic for `nextPage` was wrong (the `fetchMore()` actually returned questions array, and it wasn't empty on the last page, so it just started to reload everything from the first page again even when all 197 questions were already displayed...)

so together with fixing that logic, I also simplified it to single stable function that uses mutable refs and only creates 1 observer on mount, then disconnects the observer when there is no next page